### PR TITLE
vacuum journal remote

### LINKED ISSUE #5242: systemd-journal-remote ignores rotation rules
### Submission type

  - [x] Bug report

### systemd version the issue has been seen with

v229

### Used distribution

Ubuntu 16.04

### In case of bug report: Expected behaviour you didn't see

Remote logs were not being rotated after 2GB

### In case of bug report: Unexpected behaviour you saw

Remote logs were already being rotated every 128MB.

The server was configured with `/etc/systemd/journald.conf` containing:

```
SystemMaxFileSize=2G
RuntimeMaxFileSize=2G
```

These settings were respected by the local `systemd-journald` process, but *not* by the `systemd-journal-remote` process.
Since `man systemd-journal-remote` states that "If necessary (journal files over size, or corrupted), journal files will be rotated following normal journald rules." I would assume that it should be possible to configure the size before rotation.

### In case of bug report: Steps to reproduce the problem

1. Run `systemd-journal-remote` on one machine
2. Upload logs with `systemd-journal-upload` from one or more other machines
3. Observe that the logs will get rotated after 128MB.

### PR DESCRIPTION
**Pull request by berenddeschouwer**
This will do log rotation in split-mode none and host according to configuration set in journal-remote.conf.

Closes #5242.

### CODE REVIEW

**Comment by berenddeschouwer** [comment]
ping

**Comment by berenddeschouwer** [comment]
ping @keszybz

File: src/journal-remote/journal-remote-write.c [hunk 1]
```diff
@@ -1,6 +1,10 @@
 /* SPDX-License-Identifier: LGPL-2.1-or-later */
 
+#include <libgen.h>
+
 #include "alloc-util.h"
+#include "stat-util.h"
+#include "path-util.h"
 #include "journal-remote.h"
```
Comment (yuwata) [inline src/journal-remote/journal-remote-write.c]: Please sort the headers alphabetically.

File: src/journal-remote/journal-remote-write.c [hunk 2]
```diff
@@ -33,6 +39,17 @@ Writer* writer_new(RemoteServer *server) {
         w->n_ref = 1;
         w->server = server;
 
+        if (is_dir(server->output, true) > 0)
+                w->output = server->output;
```
Comment (yuwata) [inline src/journal-remote/journal-remote-write.c]: Please make `Writer.output` `char *`, and copy the string.

File: src/journal-remote/journal-remote-write.c [hunk 3]
```diff
@@ -33,6 +39,17 @@ Writer* writer_new(RemoteServer *server) {
...
+        else {
+                r = path_extract_directory(server->output, &dirname);
```
Comment (yuwata) [inline src/journal-remote/journal-remote-write.c]: Then, please simply pass `&w->output` here.

File: src/journal-remote/journal-remote-write.c [hunk 4]
```diff
@@ -19,12 +23,14 @@ static int do_rotate(ManagedJournalFile **f, MMapCache *m, JournalFileFlags file
...
         Writer *w;
+        _cleanup_free_ char *dirname = NULL;
+        int r;
 
         w = new0(Writer, 1);
         if (!w)
                 return NULL;
 
-        memset(&w->metrics, 0xFF, sizeof(w->metrics));
+        memcpy(&w->metrics, &server->metrics, sizeof(w->metrics));
```
Comment (yuwata) [inline src/journal-remote/journal-remote-write.c]: Please use simple copy: `w->metrics = server->metrics;`

File: src/journal-remote/journal-remote-write.c [hunk 5]
```diff
@@ -75,6 +92,9 @@ int writer_write(Writer *w,
                 r = do_rotate(&w->journal, w->mmap, file_flags);
                 if (r < 0)
                         return r;
+                r = journal_directory_vacuum(w->output, w->metrics.max_use, w->metrics.n_max_files, 0, NULL, true);
```
Comment (yuwata) [inline src/journal-remote/journal-remote-write.c]: For new code, please always comment boolean arguments with their name:
`/* verbose = */ true`

File: src/journal-remote/journal-remote-write.c [hunk 6]
```diff
@@ -93,6 +113,9 @@ int writer_write(Writer *w,
                 return r;
         else
                 log_debug("%s: Successfully rotated journal", w->journal->file->path);
+        r = journal_directory_vacuum(w->output, w->metrics.max_use, w->metrics.n_max_files, 0, NULL, true);
```
Comment (yuwata) [inline src/journal-remote/journal-remote-write.c]: Also here.

File: src/journal-remote/journal-remote-main.c [hunk 7]
```diff
@@ -53,6 +53,11 @@ static bool arg_trust_all = false;
 static bool arg_trust_all = true;
 #endif
 
+static uint64_t arg_max_use = 0;
+static uint64_t arg_max_size = 0;
+static uint64_t arg_n_max_files = 0;
+static uint64_t arg_keep_free = 0;
```
Comment (yuwata) [inline src/journal-remote/journal-remote-main.c]: As you can see the comment in journal-file.h, the default should be `UINT64_MAX`.

File: src/journal-remote/journal-remote-main.c [hunk 8]
```diff
@@ -1136,6 +1145,16 @@ static int run(int argc, char **argv) {
...
 
+        memset(&s.metrics, 0xFF, sizeof(s.metrics));
+        if (arg_max_use)
+                s.metrics.max_use = arg_max_use;
+        if (arg_max_size)
+                s.metrics.max_size = arg_max_size;
+        if (arg_keep_free)
+                s.metrics.max_size = arg_keep_free;
+        if (arg_n_max_files)
+                s.metrics.n_max_files = arg_n_max_files;
```
Comment (yuwata) [inline src/journal-remote/journal-remote-main.c]: Then, you can simply use `journal_reset_metrics()` and copy the values.

File: src/journal-remote/journal-remote.conf.in [hunk 9]
```diff
@@ -18,3 +18,7 @@
 # ServerKeyFile={{CERTIFICATE_ROOT}}/private/journal-remote.pem
 # ServerCertificateFile={{CERTIFICATE_ROOT}}/certs/journal-remote.pem
 # TrustedCertificateFile={{CERTIFICATE_ROOT}}/ca/trusted.pem
+# MaxUse=
+# KeepFree=
+# MaxFileSize=
+# MaxFiles=100
```
Comment (yuwata) [inline src/journal-remote/journal-remote.conf.in]: Hm? Where the default comes from?

File: src/journal-remote/journal-remote.h [hunk 10]
```diff
@@ -4,6 +4,7 @@
 #include "sd-event.h"
 
 #include "hashmap.h"
+#include "journal-vacuum.h"
 #include "journal-remote-parse.h"
 #include "journal-remote-write.h"
```
Comment (yuwata) [inline src/journal-remote/journal-remote.h]: Please sort the headers.

File: src/journal-remote/journal-remote-write.c [hunk 11]
```diff
@@ -33,6 +39,17 @@ Writer* writer_new(RemoteServer *server) {
         w->n_ref = 1;
         w->server = server;
 
+        if (is_dir(server->output, true) > 0)
```
Comment (yuwata) [inline src/journal-remote/journal-remote-write.c]: Here, please also comment the boolean: `/* follow = */ true`

File: man/journal-remote.conf.xml [hunk 12]
```diff
@@ -84,6 +84,16 @@
...
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>MaxUse=</varname></term>
+        <term><varname>KeepFree=</varname></term>
+        <term><varname>MaxFileSize=</varname></term>
+        <term><varname>MaxFiles=</varname></term>
+
+        <listitem><para>Enforce size limits on the journal files.
+        For more details see the documentation in journald.</para></listitem>
```
Comment (yuwata) [inline man/journal-remote.conf.xml]: Please add reference to the man page of journald.conf.
And, strictly speaking, journald.conf does not have `MaxUse=` or friends. Please mention that relevant settings, i.e. `SystemMaxUse=`.

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/journal-remote/journal-remote.conf.in (hunk 9 above)
Comment (berenddeschouwer) [reply src/journal-remote/journal-remote.conf.in]: From src/journal/journald.conf.

Ah, I see SystemMaxUse != MaxUse

File: src/journal-remote/journal-remote-write.c (hunk 1 above)
Comment (berenddeschouwer) [reply src/journal-remote/journal-remote-write.c]: done

File: src/journal-remote/journal-remote.h (hunk 10 above)
Comment (berenddeschouwer) [reply src/journal-remote/journal-remote.h]: done

File: src/journal-remote/journal-remote-write.c (hunk 11 above)
Comment (berenddeschouwer) [reply src/journal-remote/journal-remote-write.c]: done

File: src/journal-remote/journal-remote-main.c (hunk 8 above)
Comment (berenddeschouwer) [reply src/journal-remote/journal-remote-main.c]: done

File: src/journal-remote/journal-remote-write.c (hunk 5 above)
Comment (berenddeschouwer) [reply src/journal-remote/journal-remote-write.c]: done

File: src/journal-remote/journal-remote-write.c (hunk 2 above)
Comment (berenddeschouwer) [reply src/journal-remote/journal-remote-write.c]: done

File: src/journal-remote/journal-remote-write.c (hunk 3 above)
Comment (berenddeschouwer) [reply src/journal-remote/journal-remote-write.c]: done

File: src/journal-remote/journal-remote-write.c (hunk 4 above)
Comment (berenddeschouwer) [reply src/journal-remote/journal-remote-write.c]: done

File: src/journal-remote/journal-remote-write.c (hunk 6 above)
Comment (berenddeschouwer) [reply src/journal-remote/journal-remote-write.c]: done

File: src/journal-remote/journal-remote-main.c (hunk 7 above)
Comment (berenddeschouwer) [reply src/journal-remote/journal-remote-main.c]: done

File: man/journal-remote.conf.xml (hunk 12 above)
Comment (berenddeschouwer) [reply man/journal-remote.conf.xml]: I have updated the man page

**Comment by berenddeschouwer** [comment]
All the requested changes are in

File: man/journal-remote.conf.xml [hunk 13]
```diff
@@ -84,6 +84,32 @@
...
 
+      <varlistentry>
+        <term><varname>MaxUse=</varname></term>
+        <term><varname>KeepFree=</varname></term>
+        <term><varname>MaxFileSize=</varname></term>
+        <term><varname>MaxFiles=</varname></term>
+
+        <listitem><para>These are analogous to <varname>SystemMaxUse=</varname>,
+        <varname>SystemKeepFree=</varname>, <varname>SystemMaxFileSize=</varname>
+        and <varname>SystemMaxFiles=</varname> in journald.conf.</para>
```
Comment (yuwata) [inline man/journal-remote.conf.xml]: Please use `<citerefentry><refentrytitle>` tags for `journald.conf`.

File: man/journal-remote.conf.xml [hunk 14]
```diff
@@ -84,6 +84,32 @@
...
+
+        <para><varname>MaxUse=</varname> controls how much disk space
+        the systemd-journal-remote may use up at most.
```
Comment (yuwata) [inline man/journal-remote.conf.xml]: Please wrap `systemd-journal-remote` with `<command>` tag.

File: man/journal-remote.conf.xml [hunk 15]
```diff
@@ -84,6 +84,32 @@
...
+        <varname>KeepFree=</varname> controls how much disk
+        space systemd-journal-remote shall leave free for other uses.
```
Comment (yuwata) [inline man/journal-remote.conf.xml]: Also here.

File: man/journal-remote.conf.xml [hunk 16]
```diff
@@ -84,6 +84,32 @@
...
+        <command>systemd-journaldremote</command> will respect both limits
```
Comment (yuwata) [inline man/journal-remote.conf.xml]: `systemd-journaldremote` -> `systemd-journal-remote`

File: man/journal-remote.conf.xml [hunk 17]
```diff
@@ -94,6 +120,7 @@
         <citerefentry><refentrytitle>systemd-journal-remote.service</refentrytitle><manvolnum>8</manvolnum></citerefentry>,
         <citerefentry><refentrytitle>systemd</refentrytitle><manvolnum>1</manvolnum></citerefentry>,
         <citerefentry><refentrytitle>systemd-journald.service</refentrytitle><manvolnum>8</manvolnum></citerefentry>
+        <citerefentry><refentrytitle>journald.conf</refentrytitle><manvolnum>5</manvolnum></citerefentry>
```
Comment (yuwata) [inline man/journal-remote.conf.xml]: Please sort, and separate entries with `,`.

File: src/journal-remote/journal-remote-main.c [hunk 18]
```diff
@@ -1136,6 +1145,13 @@ static int run(int argc, char **argv) {
                 s.check_trust = !arg_trust_all;
         }
 
+        memset(&s.metrics, 0xFF, sizeof(s.metrics));
```
Comment (yuwata) [inline src/journal-remote/journal-remote-main.c]: This is not necessary. As `journal_reset_metrics()` does the similar.

File: src/journal-remote/journal-remote.conf.in (hunk 9 above)
Comment (yuwata) [inline src/journal-remote/journal-remote.conf.in]: Nowhere the default `100` is set in the code. I think it is better to drop it (also drop the value mentioned in the man page) to keep the backward compatibility.

**Review by yuwata — CHANGES_REQUESTED** [review]

File: man/journal-remote.conf.xml (hunk 16 above)
Comment (berenddeschouwer) [reply man/journal-remote.conf.xml]: done. doh!

File: man/journal-remote.conf.xml (hunk 13 above)
Comment (berenddeschouwer) [reply man/journal-remote.conf.xml]: ok

File: man/journal-remote.conf.xml (hunk 14 above)
Comment (berenddeschouwer) [reply man/journal-remote.conf.xml]: done

File: man/journal-remote.conf.xml (hunk 15 above)
Comment (berenddeschouwer) [reply man/journal-remote.conf.xml]: done

File: man/journal-remote.conf.xml (hunk 17 above)
Comment (berenddeschouwer) [reply man/journal-remote.conf.xml]: Is there a preference for where - goes alphabetically?

Does journal-remote come before or after journald?  ASCII rules?

File: src/journal-remote/journal-remote-main.c (hunk 18 above)
Comment (berenddeschouwer) [reply src/journal-remote/journal-remote-main.c]: done

File: src/journal-remote/journal-remote.conf.in (hunk 9 above)
Comment (berenddeschouwer) [reply src/journal-remote/journal-remote.conf.in]: I've removed it here.

File: man/journal-remote.conf.xml [hunk 19]
```diff
@@ -84,15 +84,43 @@
...
+      <varlistentry>
+        <term><varname>MaxUse=</varname></term>
+        <term><varname>KeepFree=</varname></term>
+        <term><varname>MaxFileSize=</varname></term>
+        <term><varname>MaxFiles=</varname></term>
+
+        <listitem><para>These are analogous to <varname>SystemMaxUse=</varname>,
+        <varname>SystemKeepFree=</varname>, <varname>SystemMaxFileSize=</varname>
+	and <varname>SystemMaxFiles=</varname> in
+	<citerefentry><refentrytitle>journald.conf.</refentrytitle></citerefentry></para>
```
Comment (yuwata) [inline man/journal-remote.conf.xml]: Spurious indentation. Missing manvolume tag.

File: man/journal-remote.conf.xml [hunk 20]
```diff
@@ -84,15 +84,43 @@
...
+
+        <para><varname>MaxUse=</varname> controls how much disk space
+	the <command>systemd-journal-remote</command> may use up at most.
+        <varname>KeepFree=</varname> controls how much disk
+	space <command>systemd-journal-remote</command> shall leave free for other uses.
```
Comment (yuwata) [inline man/journal-remote.conf.xml]: Spurious indentation.

File: man/journal-remote.conf.xml [hunk 21]
```diff
@@ -92,13 +92,13 @@
 
         <listitem><para>These are analogous to <varname>SystemMaxUse=</varname>,
         <varname>SystemKeepFree=</varname>, <varname>SystemMaxFileSize=</varname>
-	and <varname>SystemMaxFiles=</varname> in
-	<citerefentry><refentrytitle>journald.conf.</refentrytitle></citerefentry></para>
+        and <varname>SystemMaxFiles=</varname> in
+        <citerefentry><refentrytitle>journald.conf.</refentrytitle><manvolnum>5</manvolnum></citerefentry></para>
```
Comment (yuwata) [inline man/journal-remote.conf.xml]: Please move the period at the end after the `</citerefentry>` tag.

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/journal-remote/journal-remote-write.c [hunk 22]
```diff
@@ -33,6 +38,18 @@ Writer* writer_new(RemoteServer *server) {
...
         w->server = server;
 
+        if (is_dir(server->output, /* follow = */ true) > 0) {
+                w->output = strdup(server->output);
+                if (!w->output)
+                        return NULL;
+        } else {
+                r = path_extract_directory(server->output, &w->output);
+                if (r < 0) {
+                        log_error_errno(r, "Failed to find directory of file %s", server->output);
```
Comment (yuwata) [inline src/journal-remote/journal-remote-write.c]: Please append the error cause at the end with `"%m"`:
```
log_error_errno(r, "Failed to find directory of file \"%s\": %m", server->output);
```

**Review by yuwata — APPROVED** [review]
LGTM. Let's wait for other reviewers for a while.

File: man/journal-remote.conf.xml (hunk 17 above)
Comment (yuwata) [reply man/journal-remote.conf.xml]: We have no strict rules. I just thought the order was spurious.

**Review by bluca — APPROVED** [review]

**Pull request merged**